File: 01/test_running_average_soln.cpp

/*

CE LINK: https://godbolt.org/z/f9Pbq3c8f

*/

#include <cmath>
#include <gtest/gtest.h>
#include <numeric>
#include <vector>

#include "running_average.hpp"

TEST(RunningAverage, InitialState) {
  RunningAverage ra{};
  EXPECT_DOUBLE_EQ(ra.current_average(), 0.0);
  EXPECT_EQ(ra.current_item_count(), 0);
}

TEST(RunningAverage, SingleItem) {
  RunningAverage ra{};
  ra.add_item(5.0);
  EXPECT_DOUBLE_EQ(ra.current_average(), 5.0);
  EXPECT_EQ(ra.current_item_count(), 1);
}

TEST(RunningAverage, MultipleItems) {
  RunningAverage ra{};
  std::vector<double> items = {2.0, 1e30, 4.0, 6.0, 0};

  for (const auto &item : items) {
    ra.add_item(item);
  }
  EXPECT_EQ(ra.current_item_count(), items.size());
  double expected_average =
      std::accumulate(items.begin(), items.end(), 0.0) / items.size();
  EXPECT_DOUBLE_EQ(ra.current_average(), expected_average);
}

int main(int argc, char **argv) {
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: 01/test_running_average_soln.py
import pytest

import running_average

def test_initial_state():
    ra = running_average.RunningAverage()
    assert ra.current_item_count() == 0
    assert ra.current_average() == 0.0

def test_add_single_item():
    ra = running_average.RunningAverage()
    ra.add_item(5.0)
    assert ra.current_item_count() == 1
    assert ra.current_average() == 5.0

def test_add_multiple_items():
    ra = running_average.RunningAverage()
    items = [2.0, 1e30, 4.0, 6.0, 0]
    for item in items:
        ra.add_item(item)

    assert ra.current_item_count() == len(items)
    expected_average = sum(items) / len(items)
    assert ra.current_average() == pytest.approx(expected_average)
